client/lib/cgutil/cpuset_manager_linux.go

```diff
@@ -0,0 +1,315 @@
+}
+
+func (c *cpusetManager) RemoveAlloc(allocID string) {
+	c.mu.Lock()
+	delete(c.cgroupInfo, allocID)
+	c.mu.Unlock()
+	go c.signalReconcile()
+}
+
+func (c *cpusetManager) CgroupPathFor(allocID, task string) CgroupPathGetter {
+	return func(ctx context.Context) (string, error) {
+		c.mu.Lock()
+		allocInfo, ok := c.cgroupInfo[allocID]
+		if !ok {
+			c.mu.Unlock()
+			return "", fmt.Errorf("alloc not found for id %q", allocID)
+		}
+
+		taskInfo, ok := allocInfo[task]
+		c.mu.Unlock()
+		if !ok {
+			return "", fmt.Errorf("task %q not found", task)
+		}
+
+		for {
+			if taskInfo.Error != nil {
+				break
+			}
+			if _, err := os.Stat(taskInfo.CgroupPath); os.IsNotExist(err) {
+				select {
+				case <-ctx.Done():
+					return taskInfo.CgroupPath, ctx.Err()
+				case <-time.After(100 * time.Millisecond):
+					continue
+				}
+			}
+			break
+		}
+
+		return taskInfo.CgroupPath, taskInfo.Error
+	}
+
+}
+
+type allocTaskCgroupInfo map[string]*TaskCgroupInfo
+
+// Init checks that the cgroup parent and expected child cgroups have been created
+// If the cgroup parent is set to /nomad then this will ensure that the /nomad/shared
+// cgroup is initialized.
+func (c *cpusetManager) Init() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	cgroupParentPath, err := getCgroupPathHelper("cpuset", c.cgroupParent)
+	if err != nil {
+		return err
+	}
+	c.cgroupParentPath = cgroupParentPath
+
+	// ensures that shared cpuset exists and that the cpuset values are copied from the parent if created
+	if err := cpusetEnsureParent(filepath.Join(cgroupParentPath, SharedCpusetCgroupName)); err != nil {
+		return err
+	}
+
+	parentCpus, parentMems, err := getCpusetSubsystemSettings(cgroupParentPath)
+	if err != nil {
+		return fmt.Errorf("failed to detect parent cpuset settings: %v", err)
+	}
+	c.parentCpuset, err = cpuset.Parse(parentCpus)
+	if err != nil {
+		return fmt.Errorf("failed to parse parent cpuset.cpus setting: %v", err)
+	}
+
+	// ensure the reserved cpuset exists, but only copy the mems from the parent if creating the cgroup
+	if err := os.Mkdir(filepath.Join(cgroupParentPath, ReservedCpusetCgroupName), 0755); err == nil {
+		// cgroup created, leave cpuset.cpus empty but copy cpuset.mems from parent
+		if err != nil {
+			return err
+		}
+		if err := fscommon.WriteFile(filepath.Join(cgroupParentPath, ReservedCpusetCgroupName), "cpuset.mems", parentMems); err != nil {
+			return err
+		}
+	} else if !os.IsExist(err) {
+		return err
+	}
+
+	c.doneCh = make(chan struct{})
+	c.signalCh = make(chan struct{})
+
+	c.logger.Info("initialized cpuset cgroup manager", "parent", c.cgroupParent, "cpuset", c.parentCpuset.String())
+
+	go c.reconcileLoop()
+	return nil
+}
+
+func (c *cpusetManager) reconcileLoop() {
+	timer := time.NewTimer(0)
+	if !timer.Stop() {
+		<-timer.C
+	}
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-c.doneCh:
+			c.logger.Debug("shutting down reconcile loop")
+			return
+		case <-c.signalCh:
+			timer.Reset(500 * time.Millisecond)
```

Comment:
Just for my education, is resetting this a half-second out (rather than making it more instantaneous) a matter of avoiding blocking on this channel during a large change of allocs?

Reply:
Yeah that and batching a single reconcile loop during large alloc updates. 500ms is just a reasonable guess though.